src/transport/net_ib.cc

```diff
@@ -500,6 +532,209 @@ static_assert((offsetof(struct ncclIbRecvComm, remFifo) % 32) == 0, "ncclIbSendC
+        if (attributes.type == cudaMemoryTypeDevice)
+        {
+          host_data = malloc(r->send.size);
+          cudaMemcpy(host_data, r->send.data, r->send.size, cudaMemcpyDeviceToHost);
+        }
+        else
+        {
+          host_data = r->send.data;
+        }
+        while (offset < r->send.size) NCCLCHECK(ncclSocketProgress(NCCL_SOCKET_SEND, r->sock, host_data, r->send.size, &offset));
+        if(attributes.type == cudaMemoryTypeDevice) free(host_data);
+      }
+
+      offset = 0;
+      while (offset < sizeof(uint64_t)) NCCLCHECK(ncclSocketProgress(NCCL_SOCKET_SEND, r->sock, &r->send.wr_id, sizeof(uint64_t), &offset));
+      
+      r->events = 0;
+      break;
+    }
+    case NCCL_NET_IB_REQ_RECV:
+    {
+      if (!r->recv.fifoPosted)
+      {
+        // reset the events to correct it for socket transport.
+        r->events = r->nreqs;
+
+        while (offset < sizeof(int)) NCCLCHECK(ncclSocketProgress(NCCL_SOCKET_SEND, r->sock, &r->recv.fifoSlot, sizeof(int), &offset));
+        
+        offset = 0;
+        while (offset < sizeof(int)) NCCLCHECK(ncclSocketProgress(NCCL_SOCKET_SEND, r->sock, &r->nreqs, sizeof(int), &offset));
+        
+        offset = 0;
+        while (offset < r->nreqs * sizeof(struct ncclIbSendFifo)) NCCLCHECK(ncclSocketProgress(NCCL_SOCKET_SEND, r->sock, (void*)r->recv.fifoPtr, r->nreqs * sizeof(struct ncclIbSendFifo), &offset));
+        r->recv.fifoPosted = true;
+      }
+      break;
+    }
+    default:
+      return ncclInvalidArgument;
+  }
+  return ncclSuccess;
+}
+
+void* ncclIbSocketReceiverDaemon(void* _args) {
+  struct ncclIbRecvComm* rComm = (struct ncclIbRecvComm*)_args;
+
+  char addrline[SOCKET_NAME_MAXLEN+1];
+  INFO(NCCL_INIT|NCCL_NET, "NET/IB : Enter into ncclIbSocketReceiverDaemon addr:%s dev:%d", ncclSocketToString(&rComm->sock.addr, addrline), rComm->verbs.dev);
+
+  while(!stop.load())
+  { 
+    int slot = 0;
+    int id = 0;
+    int size = 0;
+    uint64_t wr_id = 0ULL;
+    int closed;
+    int offset = 0;
+    void* host_data = NULL;
+
+    ncclResult_t nRet = ncclSocketTryRecv(&rComm->sock, &slot, sizeof(int), &closed, false);
+    if (nRet == ncclInProgress)
+    {
+      usleep(ncclParamResilientCheckInterval());
+      continue;
+    }
+    else if (closed || nRet != ncclSuccess) break;
+    
+    disableIb(rComm->addr);
+
+    offset = 0;
+    while (offset < sizeof(int)) ncclSocketProgress(NCCL_SOCKET_RECV, &rComm->sock, &id, sizeof(int), &offset);
+    
+    offset = 0;
+    while (offset < sizeof(int)) ncclSocketProgress(NCCL_SOCKET_RECV, &rComm->sock, &size, sizeof(int), &offset);
+    
+    offset = 0;
+    if (size > 0)
+    {
+      struct cudaPointerAttributes attributes;
+      cudaError_t err = cudaPointerGetAttributes(&attributes, (void*)(uintptr_t)rComm->remFifo.elems[slot][id].addr);
+      if (err != cudaSuccess) {
+        WARN("%ld is not a valid pointer", rComm->remFifo.elems[slot][id].addr);
+        continue;
+      }
```

Comment:
just curious, if this addr is not a valid pointer, should we break the current loop and throw some errors? Why do we simply log it and continue?